tree/dataframe/src/RSqliteDS.cxx

```diff
@@ -26,6 +32,328 @@
+  return SQLITE_OK;
+}
+
+static int VfsRdOnlyLock(
+   sqlite3_file * /*pFile*/,
+   int /*level*/
+) {
+   return SQLITE_OK;
+}
+
+static int VfsRdOnlyUnlock(
+   sqlite3_file * /*pFile*/,
+   int /*level*/
+) {
+   return SQLITE_OK;
+}
+
+static int VfsRdOnlyCheckReservedLock(
+   sqlite3_file * /*pFile*/,
+   int *pResOut
+) {
+   *pResOut = 0;
+   return SQLITE_OK;
+}
+
+static int VfsRdOnlyFileControl(
+  sqlite3_file * /*p*/,
+  int /*op*/,
+  void * /*pArg*/
+) {
+  return SQLITE_NOTFOUND;
+}
+
+static int VfsRdOnlySectorSize(sqlite3_file *pFile __attribute__((unused))) {
+   return SQLITE_OPEN_READONLY;
+}
+
+static int VfsRdOnlyDeviceCharacteristics(sqlite3_file *pFile __attribute__((unused))) {
+   return SQLITE_OPEN_READONLY;
+}
+
+static int VfsRdOnlyOpen(
+   sqlite3_vfs * /*vfs*/,
+   const char *zName,
+   sqlite3_file *pFile,
+   int flags,
+   int * /*pOutFlags*/)
+{
+   VfsRootFile *p = new (pFile) VfsRootFile();
+   p->pFile.pMethods = NULL;
+
+   static const sqlite3_io_methods io_methods = {
+      1,
+      VfsRdOnlyClose,
+      VfsRdOnlyRead,
+      VfsRdOnlyWrite,
+      VfsRdOnlyTruncate,
+      VfsRdOnlySync,
+      VfsRdOnlyFileSize,
+      VfsRdOnlyLock,
+      VfsRdOnlyUnlock,
+      VfsRdOnlyCheckReservedLock,
+      VfsRdOnlyFileControl,
+      VfsRdOnlySectorSize,
+      VfsRdOnlyDeviceCharacteristics,
+      NULL,
+      NULL,
+      NULL,
+      NULL,
+      NULL,
+      NULL
+   };
+
+   if (flags & SQLITE_OPEN_READWRITE)
+      return SQLITE_IOERR;
+   if (flags & SQLITE_OPEN_DELETEONCLOSE)
+      return SQLITE_IOERR;
+   if (flags & SQLITE_OPEN_EXCLUSIVE)
+      return SQLITE_IOERR;
+
+   Davix::DavixError *err = NULL;
+   p->fd = p->pos.open(NULL, zName, O_RDONLY, &err);
+
+   if (!p->fd) {
+      printf("%s\n", err->getErrMsg().c_str());
+      return SQLITE_IOERR;
+   }
+
+   struct stat buf;
+   if (p->pos.stat(NULL, zName, &buf, NULL) == -1) {
+      return SQLITE_IOERR;
+   }
+   p->size = buf.st_size;
+
+   p->pFile.pMethods = &io_methods;
+   return 0;
+}
+
+static int VfsRdOnlyDelete(
+   sqlite3_vfs* __attribute__((unused)),
+   const char * /*zName*/,
+   int /*syncDir*/)
+{
+   return SQLITE_IOERR_DELETE;
+}
+
+static int VfsRdOnlyAccess(
+   sqlite3_vfs * /*vfs*/,
+   const char * /*zPath*/,
+   int flags,
+   int *pResOut)
+   {
+      *pResOut = 0;
+      if (flags == SQLITE_ACCESS_READWRITE) {
+         return SQLITE_OPEN_READONLY;
+      }
+   return SQLITE_OK;
+   }
+
+int VfsRdOnlyFullPathname(
+   sqlite3_vfs * /*vfs*/,
+   const char *zPath,
+   int nOut,
+   char *zOut)
+{
+   zOut[nOut-1] = '\0';
+   sqlite3_snprintf(nOut, zOut, "%s", zPath);
+   return SQLITE_OK;
+}
+
+/**
+ * Taken from unixRandomness
+ */
+static int VfsRdOnlyRandomness(
+   sqlite3_vfs * /*vfs*/,
+   int nBuf,
+   char *zBuf)
+{
+   assert((size_t)nBuf >= (sizeof(time_t) + sizeof(int)));
+   memset(zBuf, 0, nBuf);
+   pid_t randomnessPid = getpid();
+   int fd, got;
+   fd = open("/dev/urandom", O_RDONLY, 0);
```

Comment:
Can we use `uniform_int_distribution`, see e.g. https://stackoverflow.com/questions/25298585/efficiently-generating-random-bytes-of-data-in-c11-14